File: Remove Nth Node From End of List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
  //     read until the last element
  const computeDepth = (node, level) => {
    if (node && node.next) {
      return computeDepth(node.next, level + 1);
    }
    return level;
  };
  //     find the depth
  const maxDepth = computeDepth(head, 1);
  //     handle edge cases
  if (maxDepth <= 1 && n <= 1) {
    return [];
  }
  if (maxDepth === n) {
    return head.next;
  }
  //     compute the target value. target value is in front of value n. In case of the example, it's 3
  const target = maxDepth - n;
  //     finding the target, connect the next property as next.next
  let currentDepth = 1;
  const findTarget = (node, val) => {
    if (val === target) {
      node.next = node.next.next;
      return node;
    } else {
      return findTarget(node.next, val + 1);
    }
  };
  findTarget(head, 1);
  //     return head
  return head;
};

File: Leetcode_solutions/Remove_Nth_Node_From_End_of_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head
        length = 0
        first = head
        while first != None:
            length += 1
            first = first.next
        length -= n
        first = dummy
        while length > 0:
            length -= 1
            first = first.next
        first.next = first.next.next
        return dummy.next


class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy = ListNode(0)
        dummy.next = head
        first = dummy
        second = dummy
        for i in range(1, n + 2):
            first = first.next
        while first != None:
            first = first.next
            second = second.next
        second.next = second.next.next
        return dummy.next



# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        counter = 1
        current = head
        currentHead = head
        while current.next != None:
            counter += 1
            current=current.next
        targetIndex = counter - n + 1
        if targetIndex != 1:
            while targetIndex > 2:
                targetIndex -=1
                currentHead = currentHead.next
            currentHead.next = currentHead.next.next
        else:
            head=head.next
        return head